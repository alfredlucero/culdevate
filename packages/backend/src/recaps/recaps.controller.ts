import { Response } from "express";
import { RequestWithUser } from "../interfaces/requestWithUser";
import RecapsDao from "./recaps.dao";
import { Recap } from "./recaps.model";

const RecapsController = {
  async createRecap(req: RequestWithUser, res: Response) {
    const currentUser = req.user;
    const recap: Recap = {
      ...req.body,
      userId: currentUser.id,
    };

    try {
      const createdRecap = await RecapsDao.createRecap(recap);

      res.status(201).json(createdRecap);
    } catch (err) {
      res.status(500).json({ message: "Failed to create recap." });
    }
  },

  async updateRecap(req: RequestWithUser, res: Response) {
    const currentUser = req.user;
    const recapId = req.params.recapId;
    const updatedRecap: Recap = {
      ...req.body,
      userId: currentUser.id,
    };

    try {
      const foundRecap = await RecapsDao.findRecapById(recapId);
      const isMatchingUserId = foundRecap && foundRecap.userId.toString() === currentUser.id;

      if (!foundRecap || !isMatchingUserId) {
        return res.status(404).json({ message: "Failed to find matching recap to update" });
      }

      const changedRecap = await RecapsDao.updateRecapById({ recapId, updatedRecap });

      if (changedRecap) {
        res.status(200).json(changedRecap);
      } else {
        // If by some chance it got deleted while we were trying to update it
        // we'll still get a 404 back
        res.status(404).json({ message: "Failed to find matching recap to update" });
      }
    } catch (err) {
      res.status(500).json({ message: "Failed to update recap" });
    }
  },

  async deleteRecap(req: RequestWithUser, res: Response) {
    const recapId = req.params.recapId;
    const currentUser = req.user;

    try {
      const foundRecap = await RecapsDao.findRecapById(recapId);
      const isMatchingUserId = foundRecap && foundRecap.userId.toString() === currentUser.id;

      if (!foundRecap || !isMatchingUserId) {
        return res.status(404).json({ message: "Failed to find matching recap to delete" });
      }

      const deletedRecap = await RecapsDao.removeRecapById(recapId);

      if (deletedRecap) {
        res.status(200).json(deletedRecap);
      } else {
        // If by some chance it got deleted while we were trying to delete it
        // we'll still get a 404 back
        res.status(404).json({ message: "Failed to find matching recap to delete" });
      }
    } catch (err) {
      res.status(500).json({ message: "Failed to delete recap" });
    }
  },
};

export default RecapsController;
