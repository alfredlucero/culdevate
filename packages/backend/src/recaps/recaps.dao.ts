import {
  RecapBaseModel,
  RecapWorkExperienceModel,
  RecapEducationModel,
  RecapAccomplishmentsModel,
  RecapPublicationsModel,
  RecapSkillsModel,
  RecapSideProjectsModel,
  RecapOrganizationsModel,
  RecapReferencesModel,
  RecapOtherModel,
  Recap,
} from "./recaps.model";

const RecapsDao = {
  findAllRecaps(userId: string) {
    return RecapBaseModel.find({ userId });
  },

  findRecapById(recapId: string) {
    return RecapBaseModel.findById(recapId);
  },

  findRecapByIdAndUserId({ recapId, userId }: { recapId: string; userId: string }) {
    return RecapBaseModel.findOne({ userId, _id: recapId });
  },

  createRecap(recap: Recap) {
    switch (recap.kind) {
      case "Work Experience":
        const workExperienceRecap = new RecapWorkExperienceModel(recap);
        return workExperienceRecap.save();
      case "Education":
        const educationRecap = new RecapEducationModel(recap);
        return educationRecap.save();
      case "Accomplishments":
        const accomplishmentsRecap = new RecapAccomplishmentsModel(recap);
        return accomplishmentsRecap.save();
      case "Publications":
        const publicationsRecap = new RecapPublicationsModel(recap);
        return publicationsRecap.save();
      case "Skills":
        const skillsRecap = new RecapSkillsModel(recap);
        return skillsRecap.save();
      case "Side Projects":
        const sideProjectsRecap = new RecapSideProjectsModel(recap);
        return sideProjectsRecap.save();
      case "Organizations":
        const organizationsRecap = new RecapOrganizationsModel(recap);
        return organizationsRecap.save();
      case "References":
        const referencesRecap = new RecapReferencesModel(recap);
        return referencesRecap.save();
      case "Other":
        const otherRecap = new RecapOtherModel(recap);
        return otherRecap.save();
      default:
        throw new Error("Failed to find matching recap kind to create!");
    }
  },

  updateRecapById({ recapId, updatedRecap }: { recapId: string; updatedRecap: Recap }) {
    switch (updatedRecap.kind) {
      case "Work Experience":
        return RecapWorkExperienceModel.findOneAndUpdate(
          { _id: recapId },
          { $set: updatedRecap, ...(!updatedRecap.endDate ? { $unset: { endDate: 1 } } : {}) },
          { new: true },
        );
      case "Education":
        return RecapEducationModel.findOneAndUpdate(
          { _id: recapId },
          { $set: updatedRecap, ...(!updatedRecap.endDate ? { $unset: { endDate: 1 } } : {}) },
          { new: true },
        );
      case "Accomplishments":
        return RecapAccomplishmentsModel.findOneAndUpdate({ _id: recapId }, updatedRecap, { new: true });
      case "Publications":
        return RecapPublicationsModel.findOneAndUpdate({ _id: recapId }, updatedRecap, { new: true });
      case "Skills":
        return RecapSkillsModel.findOneAndUpdate({ _id: recapId }, updatedRecap, { new: true });
      case "Side Projects":
        return RecapSideProjectsModel.findOneAndUpdate(
          { _id: recapId },
          { $set: updatedRecap, ...(!updatedRecap.endDate ? { $unset: { endDate: 1 } } : {}) },
          { new: true },
        );
      case "Organizations":
        return RecapOrganizationsModel.findOneAndUpdate(
          { _id: recapId },
          { $set: updatedRecap, ...(!updatedRecap.endDate ? { $unset: { endDate: 1 } } : {}) },
          { new: true },
        );
      case "References":
        return RecapReferencesModel.findOneAndUpdate({ _id: recapId }, updatedRecap, { new: true });
      case "Other":
        return RecapOtherModel.findOneAndUpdate({ _id: recapId }, updatedRecap, { new: true });
      default:
        throw new Error("Failed to find matching recap kind to update!");
    }
  },

  removeRecapById(recapId: string) {
    return RecapBaseModel.findByIdAndDelete({ _id: recapId });
  },
};

export default RecapsDao;
