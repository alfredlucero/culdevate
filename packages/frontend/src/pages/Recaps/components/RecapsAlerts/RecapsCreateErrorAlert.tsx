import React from "react";
import Alert, { AlertProps } from "../../../../components/Alert";
import { CommonProps } from "../../../../components/commonProps";
import { RecapKind } from "../../recaps.interface";

interface RecapsCreateErrorAlertProps extends CommonProps {
  kind: RecapKind;
  isShowing: AlertProps["isShowing"];
  onHide: AlertProps["onHide"];
}

const RecapsCreateErrorAlert: React.FC<RecapsCreateErrorAlertProps> = ({
  kind,
  isShowing,
  onHide,
  testId = "",
  className = "",
  ...passThroughProps
}) => {
  return (
    <Alert
      variant="danger"
      isShowing={isShowing}
      onHide={onHide}
      testId={testId}
      className={className}
      {...passThroughProps}
    >
      Something went wrong with adding a {kind} Recap! Please try again or let us know if there is an issue!
    </Alert>
  );
};

export default RecapsCreateErrorAlert;
