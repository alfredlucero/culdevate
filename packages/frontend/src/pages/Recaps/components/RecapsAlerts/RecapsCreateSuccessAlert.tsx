import React from "react";
import Alert, { AlertProps } from "../../../../components/Alert";
import { CommonProps } from "../../../../components/commonProps";
import { RecapKind } from "../../recaps.interface";

interface RecapsCreateSuccessAlertProps extends CommonProps {
  kind: RecapKind;
  isShowing: AlertProps["isShowing"];
  onHide: AlertProps["onHide"];
}

const RecapsCreateSuccessAlert: React.FC<RecapsCreateSuccessAlertProps> = ({
  kind,
  isShowing,
  onHide,
  testId = "",
  className = "",
  ...passThroughProps
}) => {
  return (
    <Alert
      variant="success"
      isShowing={isShowing}
      onHide={onHide}
      testId={testId}
      className={className}
      {...passThroughProps}
    >
      {kind !== RecapKind.AllRecaps && <>You have successfully created a {kind} Recap!</>}
      {kind === RecapKind.AllRecaps && <>You have successfully created a Recap!</>}
    </Alert>
  );
};

export default RecapsCreateSuccessAlert;
