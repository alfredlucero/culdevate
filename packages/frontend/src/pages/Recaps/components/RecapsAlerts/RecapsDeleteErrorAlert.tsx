import React from "react";
import Alert, { AlertProps } from "../../../../components/Alert";
import { CommonProps } from "../../../../components/commonProps";
import { RecapKind } from "../../recaps.interface";

interface RecapsDeleteErrorAlertProps extends CommonProps {
  kind: RecapKind;
  isShowing: AlertProps["isShowing"];
  onHide: AlertProps["onHide"];
}

const RecapsDeleteErrorAlert: React.FC<RecapsDeleteErrorAlertProps> = ({
  kind,
  isShowing,
  onHide,
  testId = "",
  className = "",
  ...passThroughProps
}) => {
  return (
    <Alert
      variant="danger"
      isShowing={isShowing}
      onHide={onHide}
      testId={testId}
      className={className}
      {...passThroughProps}
    >
      Something went wrong with deleting the {kind} Recap! Please try again or let us know if there is an issue!
    </Alert>
  );
};

export default RecapsDeleteErrorAlert;
