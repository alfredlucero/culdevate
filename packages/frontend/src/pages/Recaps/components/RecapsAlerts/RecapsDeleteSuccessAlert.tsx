import React from "react";
import Alert, { AlertProps } from "../../../../components/Alert";
import { CommonProps } from "../../../../components/commonProps";
import { RecapKind } from "../../recaps.interface";

interface RecapsDeleteSuccessAlertProps extends CommonProps {
  kind: RecapKind;
  isShowing: AlertProps["isShowing"];
  onHide: AlertProps["onHide"];
}

const RecapsDeleteSuccessAlert: React.FC<RecapsDeleteSuccessAlertProps> = ({
  kind,
  isShowing,
  onHide,
  testId = "",
  className = "",
  ...passThroughProps
}) => {
  return (
    <Alert
      variant="success"
      isShowing={isShowing}
      onHide={onHide}
      testId={testId}
      className={className}
      {...passThroughProps}
    >
      You have successfully deleted the {kind} Recap!
    </Alert>
  );
};

export default RecapsDeleteSuccessAlert;
