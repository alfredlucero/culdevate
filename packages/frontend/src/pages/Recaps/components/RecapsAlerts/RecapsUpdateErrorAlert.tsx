import React from "react";
import Alert, { AlertProps } from "../../../../components/Alert";
import { CommonProps } from "../../../../components/commonProps";
import { RecapKind } from "../../recaps.interface";

interface RecapsUpdateErrorAlertProps extends CommonProps {
  kind: RecapKind;
  isShowing: AlertProps["isShowing"];
  onHide: AlertProps["onHide"];
}

const RecapsUpdateErrorAlert: React.FC<RecapsUpdateErrorAlertProps> = ({
  kind,
  isShowing,
  onHide,
  testId = "",
  className = "",
  ...passThroughProps
}) => {
  return (
    <Alert
      variant="danger"
      isShowing={isShowing}
      onHide={onHide}
      testId={testId}
      className={className}
      {...passThroughProps}
    >
      Something went wrong with updating a {kind} Recap! Please try again or let us know if there is an issue!
    </Alert>
  );
};

export default RecapsUpdateErrorAlert;
