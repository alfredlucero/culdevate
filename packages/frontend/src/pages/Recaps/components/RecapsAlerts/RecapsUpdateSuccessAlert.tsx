import React from "react";
import Alert, { AlertProps } from "../../../../components/Alert";
import { CommonProps } from "../../../../components/commonProps";
import { RecapKind } from "../../recaps.interface";

interface RecapsUpdateSuccessAlertProps extends CommonProps {
  kind: RecapKind;
  isShowing: AlertProps["isShowing"];
  onHide: AlertProps["onHide"];
}

const RecapsUpdateSuccessAlert: React.FC<RecapsUpdateSuccessAlertProps> = ({
  kind,
  isShowing,
  onHide,
  testId = "",
  className = "",
  ...passThroughProps
}) => {
  return (
    <Alert
      variant="success"
      isShowing={isShowing}
      onHide={onHide}
      testId={testId}
      className={className}
      {...passThroughProps}
    >
      You have successfully updated a {kind} Recap!
    </Alert>
  );
};

export default RecapsUpdateSuccessAlert;
