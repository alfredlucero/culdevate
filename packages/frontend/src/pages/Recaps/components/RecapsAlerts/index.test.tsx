import React from "react";
import { render } from "@testing-library/react";
import RecapsCreateSuccessAlert from "./RecapsCreateSuccessAlert";
import RecapsCreateErrorAlert from "./RecapsCreateErrorAlert";
import RecapsUpdateSuccessAlert from "./RecapsUpdateSuccessAlert";
import RecapsUpdateErrorAlert from "./RecapsUpdateErrorAlert";
import RecapsDeleteSuccessAlert from "./RecapsDeleteSuccessAlert";
import RecapsDeleteErrorAlert from "./RecapsDeleteErrorAlert";
import { RecapKind } from "../../recaps.interface";

describe("<RecapAlerts />", () => {
  test("should render all alerts without error", () => {
    const { container } = render(
      <div>
        <RecapsCreateSuccessAlert
          kind={RecapKind.Education}
          isShowing={true}
          onHide={() => {}}
          className="extra-class"
          testId="testId"
        />
        <RecapsCreateErrorAlert
          kind={RecapKind.Education}
          isShowing={true}
          onHide={() => {}}
          className="extra-class"
          testId="testId"
        />
        <RecapsDeleteSuccessAlert
          kind={RecapKind.Education}
          isShowing={true}
          onHide={() => {}}
          className="extra-class"
          testId="testId"
        />
        <RecapsDeleteErrorAlert
          kind={RecapKind.Education}
          isShowing={true}
          onHide={() => {}}
          className="extra-class"
          testId="testId"
        />
        <RecapsUpdateSuccessAlert
          kind={RecapKind.Education}
          isShowing={true}
          onHide={() => {}}
          className="extra-class"
          testId="testId"
        />
        <RecapsUpdateErrorAlert
          kind={RecapKind.Education}
          isShowing={true}
          onHide={() => {}}
          className="extra-class"
          testId="testId"
        />
      </div>,
    );

    expect(container).toMatchSnapshot();
  });
});
