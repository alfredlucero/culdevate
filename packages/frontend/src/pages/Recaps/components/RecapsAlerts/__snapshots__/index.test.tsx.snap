// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecapAlerts /> should render all alerts without error 1`] = `
<div>
  <div>
    <div
      class="border-l-4 p-4 cursor-pointer rounded shadow flex justify-between break-words text-green-700 bg-green-100 border-green-500 extra-class"
      data-testid="testId"
      role="alert"
    >
      <div
        class="w-5/6"
      >
        You have successfully created a 
        Education
         Recap!
      </div>
      <div
        class="w-1/6 flex justify-end items-center"
      >
        <svg
          class="fill-current h-6 w-6 text-green-700"
          role="button"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Close
          </title>
          <path
            d="M14.348 14.849a1.2 1.2 0 0 1-1.697 0L10 11.819l-2.651 3.029a1.2 1.2 0 1 1-1.697-1.697l2.758-3.15-2.759-3.152a1.2 1.2 0 1 1 1.697-1.697L10 8.183l2.651-3.031a1.2 1.2 0 1 1 1.697 1.697l-2.758 3.152 2.758 3.15a1.2 1.2 0 0 1 0 1.698z"
          />
        </svg>
      </div>
    </div>
    <div
      class="border-l-4 p-4 cursor-pointer rounded shadow flex justify-between break-words text-red-700 bg-red-100 border-red-500 extra-class"
      data-testid="testId"
      role="alert"
    >
      <div
        class="w-5/6"
      >
        Something went wrong with adding a 
        Education
         Recap! Please try again or let us know if there is an issue!
      </div>
      <div
        class="w-1/6 flex justify-end items-center"
      >
        <svg
          class="fill-current h-6 w-6 text-red-700"
          role="button"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Close
          </title>
          <path
            d="M14.348 14.849a1.2 1.2 0 0 1-1.697 0L10 11.819l-2.651 3.029a1.2 1.2 0 1 1-1.697-1.697l2.758-3.15-2.759-3.152a1.2 1.2 0 1 1 1.697-1.697L10 8.183l2.651-3.031a1.2 1.2 0 1 1 1.697 1.697l-2.758 3.152 2.758 3.15a1.2 1.2 0 0 1 0 1.698z"
          />
        </svg>
      </div>
    </div>
    <div
      class="border-l-4 p-4 cursor-pointer rounded shadow flex justify-between break-words text-green-700 bg-green-100 border-green-500 extra-class"
      data-testid="testId"
      role="alert"
    >
      <div
        class="w-5/6"
      >
        You have successfully deleted the 
        Education
         Recap!
      </div>
      <div
        class="w-1/6 flex justify-end items-center"
      >
        <svg
          class="fill-current h-6 w-6 text-green-700"
          role="button"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Close
          </title>
          <path
            d="M14.348 14.849a1.2 1.2 0 0 1-1.697 0L10 11.819l-2.651 3.029a1.2 1.2 0 1 1-1.697-1.697l2.758-3.15-2.759-3.152a1.2 1.2 0 1 1 1.697-1.697L10 8.183l2.651-3.031a1.2 1.2 0 1 1 1.697 1.697l-2.758 3.152 2.758 3.15a1.2 1.2 0 0 1 0 1.698z"
          />
        </svg>
      </div>
    </div>
    <div
      class="border-l-4 p-4 cursor-pointer rounded shadow flex justify-between break-words text-red-700 bg-red-100 border-red-500 extra-class"
      data-testid="testId"
      role="alert"
    >
      <div
        class="w-5/6"
      >
        Something went wrong with deleting the 
        Education
         Recap! Please try again or let us know if there is an issue!
      </div>
      <div
        class="w-1/6 flex justify-end items-center"
      >
        <svg
          class="fill-current h-6 w-6 text-red-700"
          role="button"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Close
          </title>
          <path
            d="M14.348 14.849a1.2 1.2 0 0 1-1.697 0L10 11.819l-2.651 3.029a1.2 1.2 0 1 1-1.697-1.697l2.758-3.15-2.759-3.152a1.2 1.2 0 1 1 1.697-1.697L10 8.183l2.651-3.031a1.2 1.2 0 1 1 1.697 1.697l-2.758 3.152 2.758 3.15a1.2 1.2 0 0 1 0 1.698z"
          />
        </svg>
      </div>
    </div>
    <div
      class="border-l-4 p-4 cursor-pointer rounded shadow flex justify-between break-words text-green-700 bg-green-100 border-green-500 extra-class"
      data-testid="testId"
      role="alert"
    >
      <div
        class="w-5/6"
      >
        You have successfully updated a 
        Education
         Recap!
      </div>
      <div
        class="w-1/6 flex justify-end items-center"
      >
        <svg
          class="fill-current h-6 w-6 text-green-700"
          role="button"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Close
          </title>
          <path
            d="M14.348 14.849a1.2 1.2 0 0 1-1.697 0L10 11.819l-2.651 3.029a1.2 1.2 0 1 1-1.697-1.697l2.758-3.15-2.759-3.152a1.2 1.2 0 1 1 1.697-1.697L10 8.183l2.651-3.031a1.2 1.2 0 1 1 1.697 1.697l-2.758 3.152 2.758 3.15a1.2 1.2 0 0 1 0 1.698z"
          />
        </svg>
      </div>
    </div>
    <div
      class="border-l-4 p-4 cursor-pointer rounded shadow flex justify-between break-words text-red-700 bg-red-100 border-red-500 extra-class"
      data-testid="testId"
      role="alert"
    >
      <div
        class="w-5/6"
      >
        Something went wrong with updating a 
        Education
         Recap! Please try again or let us know if there is an issue!
      </div>
      <div
        class="w-1/6 flex justify-end items-center"
      >
        <svg
          class="fill-current h-6 w-6 text-red-700"
          role="button"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Close
          </title>
          <path
            d="M14.348 14.849a1.2 1.2 0 0 1-1.697 0L10 11.819l-2.651 3.029a1.2 1.2 0 1 1-1.697-1.697l2.758-3.15-2.759-3.152a1.2 1.2 0 1 1 1.697-1.697L10 8.183l2.651-3.031a1.2 1.2 0 1 1 1.697 1.697l-2.758 3.152 2.758 3.15a1.2 1.2 0 0 1 0 1.698z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
